src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -153,8 +154,71 @@ class FastSync(
       assignedHandlers -= handler
     }
 
+    private def insertHeader(header: BlockHeader): Boolean = {
+      val parentTdOpt = blockchain.getTotalDifficultyByHash(header.parentHash)
+      parentTdOpt foreach { parentTotalDifficulty =>
+        blockchain.save(header)
+        blockchain.save(header.hash, parentTotalDifficulty + header.difficulty)
+
+        if (header.number > syncState.bestBlockHeaderNumber) {
+          syncState = syncState.copy(bestBlockHeaderNumber = header.number)
+        }
+
+        syncState = syncState
+          .enqueueBlockBodies(Seq(header.hash))
+          .enqueueReceipts(Seq(header.hash))
+      }
+
+      parentTdOpt.isDefined
+    }
+
+    @tailrec
+    private def processHeaders(peer: Peer, headers: Seq[BlockHeader]): Unit = {
+      if (headers.nonEmpty) {
+        val header = headers.head
+        val remaining = headers.tail
+
+        val shouldValidate = header.number >= syncState.nextBlockToFullyValidate
+        if (shouldValidate) {
+          syncState = syncState.copy(
+            nextBlockToFullyValidate =
+              if (syncState.bestBlockHeaderNumber >= syncState.targetBlock.number - X) header.number + 1
+              else header.number + K / 2 + Random.nextInt(K))
+
+          validators.blockHeaderValidator.validate(header, blockchain) match {
+            case Right(_) =>
+              if (insertHeader(header)) processHeaders(peer, remaining)
+              else ()
+
+            case Left(error) =>
+              log.warning(s"Block header validation failed during fast sync at block ${header.number}: $error")
+              blacklist(peer.id, blacklistDuration, "block header validation failed")
+
+              // discard last N blocks
```

Comment:
I don't quite follow this. If we found that a specific block is invalid, why previous `N` ones are dropped? Those might be valid, or not?

Reply:
Since we only validate every ~Xth block (+ some randomization) we cannot be sure these are valid blocks.
See https://github.com/ethereum/go-ethereum/pull/1889 for comprehensive explanation. In short:
> With this caveat calculated, the fast sync should be modified so that up to the pivoting point - X, only every K=100-th header should be verified (at random), after which all headers up to pivot point + X should be fully verified before starting state database downloading. Note: if a sync fails due to header verification the last N headers must be discarded as they cannot be trusted enough.

Reply:
Should it always be `N` when `N` are available?

In other words: forget the begining of the chain, I'm talking about a situation when we are in the `[target - X, target]` range. Shouldn't it be:

```
val numberOfBlocksToDiscard = if (target - current <= X) N + X - (target - current) else N
```
(maths may need double-checking)

So, `N` plus the number of blocks already validated in the `X` phase?

Reply:
Why do you think we should drop the extra blocks? For me it doesn't seem to make a difference in this context. (is there a difference if validation fails at block target/2 or target-10? I think it's sufficient to drop N in both cases)

Reply:
Somehow I thought we should revert to a fully validated block, but since `N` is not divisible by `K` that is not the case. Anyway, the way put it - no, I don't think it will make a difference 👍